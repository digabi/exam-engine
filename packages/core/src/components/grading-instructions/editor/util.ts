import { DOMSerializer, Fragment, Schema } from 'prosemirror-model'

export const serializeFragment = (schema: Schema, fragment: Fragment) => {
  const serializedFragment = DOMSerializer.fromSchema(schema).serializeFragment(fragment, {
    document: document
  })

  const transformNode = (node: Node) => {
    if (node.nodeType === Node.ELEMENT_NODE) {
      const element = node as HTMLElement

      if (element.nodeName === 'IMG' && element.classList.contains('e-image')) {
        const eImage = document.createElement('e:image')
        eImage.innerHTML = element.innerHTML

        const src = element.getAttribute('src') || ''
        const filename = src.split('/').pop() || ''

        Array.from(element.attributes).forEach(attr => {
          eImage.setAttribute(attr.name, attr.name === 'src' ? filename : attr.value)
        })

        element.replaceWith(eImage)
      }
    }

    node.childNodes.forEach(transformNode)
  }

  serializedFragment.childNodes.forEach(transformNode)

  return serializedFragment
}
