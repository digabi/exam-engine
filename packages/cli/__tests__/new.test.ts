import * as fs from 'fs/promises'
import { exec, stripColorCodes } from './util'
import path from 'path'

describe('ee new', () => {
  const cwd = path.resolve(__dirname, '../../..')
  let output: string

  beforeAll(removeDirectory)
  beforeAll(async () => {
    output = await exec('yarn ee new new_exam', { cwd })
  })
  afterAll(removeDirectory)

  it('logs exam creation', () => {
    expect(stripColorCodes(output)).toContain(`âœ” new_exam`)
  })

  it('creates new exam skeleton', async () => {
    const list = await fs.readdir(path.resolve(cwd, 'new_exam'))
    const newExam = await fs.readFile(path.resolve(cwd, 'new_exam', 'exam.xml'), { encoding: 'utf8' })
    expect(list).toEqual(['attachments', 'exam.xml'])
    expect(newExam).toMatchSnapshot()
  })

  function removeDirectory() {
    return fs.rm(path.resolve(cwd, 'new_exam'), { recursive: true, force: true })
  }
})
