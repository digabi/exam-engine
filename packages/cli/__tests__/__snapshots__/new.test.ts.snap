// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ee new creates new exam skeleton 1`] = `
"<?xml version="1.0" encoding="utf-8" ?>
<e:exam xmlns:e="http://ylioppilastutkinto.fi/exam.xsd" xmlns="http://www.w3.org/1999/xhtml" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://ylioppilastutkinto.fi/exam.xsd https://abitti.dev/schema/exam.xsd" exam-schema-version="0.5">
    <e:exam-versions>
        <e:exam-version lang="fi-FI" />
    </e:exam-versions>

    <e:exam-title>Exam title</e:exam-title>

    <e:exam-instruction>
        Exam instructions
    </e:exam-instruction>

    <e:table-of-contents />

    <e:section>
        <e:section-title>Section title</e:section-title>

        <e:question>
            <e:question-title>Question title</e:question-title>
            <e:question-instruction>Question instructions</e:question-instruction>
            <e:text-answer max-score="10" type="rich-text" />
        </e:question>
    </e:section>
</e:exam>
"
`;
